classes/kohana/encrypt.php

```diff
@@ -1,213 +1,134 @@
- *
+ * using the [Sodium](https://www.php.net/manual/en/book.sodium.php) extension.
+
  * @package    Kohana
  * @category   Security
  * @author     Kohana Team
  * @copyright  (c) 2007-2012 Kohana Team
  * @license    http://kohanaframework.org/license
  */
 class Kohana_Encrypt {
-
 	/**
-	 * @var  string  default instance name
+	 * @var string
 	 */
 	public static $default = 'default';
 
 	/**
-	 * @var  array  Encrypt class instances
+	 * @var array
 	 */
-	public static $instances = array();
+	public static $instances = [];
 
 	/**
-	 * @var  string  OS-dependent RAND type to use
+	 * @var string
 	 */
-	protected static $_rand;
+	private $key;
 
 	/**
 	 * Returns a singleton instance of Encrypt. An encryption key must be
 	 * provided in your "encrypt" configuration file.
 	 *
-	 *     $encrypt = Encrypt::instance();
+	 * @param string $name configuration group name
 	 *
-	 * @param   string  $name   configuration group name
-	 * @return  Encrypt
+	 * @return self
 	 */
-	public static function instance($name = NULL)
+	public static function instance($name = null)
 	{
-		if ($name === NULL)
+		if ($name === null)
 		{
-			// Use the default instance name
-			$name = Encrypt::$default;
+			$name = self::$default;
 		}
 
-		if ( ! isset(Encrypt::$instances[$name]))
+		if (!isset(self::$instances[$name]))
 		{
 			// Load the configuration data
-			$config = Kohana::$config->load('encrypt')->$name;
+			$config = \Kohana::$config->load('encrypt')->$name;
 
-			if ( ! isset($config['key']))
+			if (!isset($config['key']))
 			{
 				// No default encryption key is provided!
-				throw new Kohana_Exception('No encryption key is defined in the encryption configuration group: :group',
+				throw new \Kohana_Exception('No encryption key is defined in the encryption configuration group: :group',
 					array(':group' => $name));
 			}
 
-			if ( ! isset($config['mode']))
-			{
-				// Add the default mode
-				$config['mode'] = MCRYPT_MODE_NOFB;
-			}
-
-			if ( ! isset($config['cipher']))
-			{
-				// Add the default cipher
-				$config['cipher'] = MCRYPT_RIJNDAEL_128;
-			}
-
 			// Create a new instance
-			Encrypt::$instances[$name] = new Encrypt($config['key'], $config['mode'], $config['cipher']);
+			self::$instances[$name] = new self($config['key']);
 		}
 
-		return Encrypt::$instances[$name];
+		return self::$instances[$name];
 	}
 
 	/**
-	 * Creates a new mcrypt wrapper.
-	 *
-	 * @param   string  $key    encryption key
-	 * @param   string  $mode   mcrypt mode
-	 * @param   string  $cipher mcrypt cipher
+	 * @param string $key Encryption key (hexadecimal)
 	 */
-	public function __construct($key, $mode, $cipher)
+	public function __construct(string $key)
 	{
-		// Find the max length of the key, based on cipher and mode
-		$size = mcrypt_get_key_size($cipher, $mode);
+		$key = hex2bin($key);
 
-		if (isset($key[$size]))
+		if (mb_strlen($key, '8bit') !== SODIUM_CRYPTO_SECRETBOX_KEYBYTES)
```

Comment:
Just use normal `strlen()` to keep in common with the rest of the code.

```suggestion
		if (strlen($key) !== SODIUM_CRYPTO_SECRETBOX_KEYBYTES)
```